applications/nrf_desktop/src/hw_interface/click.c

```diff
@@ -0,0 +1,234 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <zephyr.h>
+
+#define MODULE click
+#include "module_state_event.h"
+#include "button_event.h"
+#include "click_event.h"
+
+#include "click_def.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(MODULE, CONFIG_DESKTOP_CLICK_LOG_LEVEL);
+
+#define SHORT_CLICK_MAX		K_MSEC(500)
+#define LONG_CLICK_MIN		K_MSEC(5000)
+#define CLICK_ON_INIT_MAX	K_MSEC(500)
+#define HOLD_ON_INIT_MIN	K_MSEC(5000)
+
+enum state {
+	STATE_DISABLED,
+	STATE_INIT_CLICK_CHECK,
+	STATE_INIT_HOLD_CHECK,
+	STATE_ACTIVE,
+};
+
+struct click_detector {
+	const struct click_detector_config *config;
+	struct k_delayed_work single_click;
+	struct k_delayed_work long_click;
+	bool button_pressed_on_init;
+	u32_t timestamp;
+};
+
+static struct click_detector detectors[ARRAY_SIZE(click_configs)];
+static enum state state;
+
+struct k_delayed_work click_on_init;
+
+static void submit_click_event(struct click_detector *detector,
+			       enum click click)
+{
+	struct click_event *event = new_click_event();
+
+	event->detector_id = detector->config->detector_id;
+	event->click = click;
+
+	EVENT_SUBMIT(event);
+}
+
+static void single_click_handler(struct k_work *work)
+{
+	struct click_detector *detector = CONTAINER_OF(work,
+						       struct click_detector,
+						       single_click);
+	submit_click_event(detector, CLICK_SHORT);
+
+	/* Reset workqueue to make sure cancel will notice work was done. */
+	detector->single_click.work_q = NULL;
+}
+
+static void long_click_handler(struct k_work *work)
+{
+	struct click_detector *detector = CONTAINER_OF(work,
+						       struct click_detector,
+						       long_click);
+	detector->timestamp = 0;
+	submit_click_event(detector, CLICK_LONG);
+
+	/* Reset workqueue to make sure cancel will notice work was done. */
+	detector->single_click.work_q = NULL;
```

Comment:
why do you clear single_click member here?

Reply:
This will be removed (when starting using single work)

Reply:
@ppryga this abuse of API was added to enforce the return value from a cancel call later on.

Reply:
I was wondering why single_click.work_q is set to NULL in long_clik work, is that correct?